substrate/primitives/runtime/src/proving_trie.rs

```diff
@@ -0,0 +1,235 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Types for a simple merkle trie used for checking and generating proofs.
+
+use crate::{Decode, DispatchError, Encode};
+
+use sp_std::vec::Vec;
+use sp_trie::{
+	trie_types::{TrieDBBuilder, TrieDBMutBuilderV1},
+	LayoutV1, MemoryDB, Recorder, Trie, TrieMut, EMPTY_PREFIX,
+};
+
+type HashOf<Hashing> = <Hashing as sp_core::Hasher>::Out;
+
+/// A basic trie implementation for checking and generating proofs for a key / value pair.
+pub struct BasicProvingTrie<Hashing, Key, Value>
+where
+	Hashing: sp_core::Hasher,
+{
+	db: MemoryDB<Hashing>,
+	root: HashOf<Hashing>,
+	_phantom: core::marker::PhantomData<(Key, Value)>,
+}
+
+impl<Hashing, Key, Value> BasicProvingTrie<Hashing, Key, Value>
+where
+	Hashing: sp_core::Hasher,
+	Key: Encode,
+	Value: Encode + Decode,
+{
+	/// Create a new instance of a `ProvingTrie` using an iterator of key/value pairs.
+	pub fn generate_for<I>(items: I) -> Result<Self, DispatchError>
+	where
+		I: IntoIterator<Item = (Key, Value)>,
+	{
+		let mut db = MemoryDB::default();
+		let mut root = Default::default();
+
+		{
+			let mut trie = TrieDBMutBuilderV1::new(&mut db, &mut root).build();
+			for (key, value) in items.into_iter() {
+				key.using_encoded(|k| value.using_encoded(|v| trie.insert(k, v)))
+					.map_err(|_| "failed to insert into trie")?;
+			}
+		}
+
+		Ok(Self { db, root, _phantom: Default::default() })
+	}
+
+	/// Access the underlying trie root.
+	pub fn root(&self) -> &HashOf<Hashing> {
+		&self.root
+	}
+
+	/// Check a proof contained within the current `MemoryDB`. Returns `None` if the
+	/// nodes within the current `MemoryDB` are insufficient to query the item.
+	pub fn query(&self, key: Key) -> Option<Value> {
+		let trie = TrieDBBuilder::new(&self.db, &self.root).build();
+		key.using_encoded(|s| trie.get(s))
+			.ok()?
+			.and_then(|raw| Value::decode(&mut &*raw).ok())
+	}
+
+	/// Create the full verification data needed to prove all `keys` and their values in the trie.
+	/// Returns `None` if the nodes within the current `MemoryDB` are insufficient to create a
+	/// proof.
```

Comment:
Since we do not pass the trie version in parameter and use v1 as default, it seems important to me to state it in the doc.
``` Proof are created with latest substrate trie format, runtime using non latest should not use it (eg trie_version = 0).```

Reply:
done